jupyterhub/tests/test_auth.py

```diff
@@ -591,3 +591,79 @@ async def test_auth_managed_groups(
         assert not app.db.dirty
         groups = sorted(g.name for g in user.groups)
         assert groups == expected_refresh_groups
+
+
+def getRoleNames(role_list):
+    return [role['name'] for role in role_list]
+
+
+class MockRolesAuthenticator(auth.Authenticator):
+    authenticated_roles = Any()
+    refresh_roles = Any()
+    manage_roles = True
+
+    def authenticate(self, handler, data):
+        return {
+            "name": data["username"],
+            "roles": self.authenticated_roles,
+        }
+
+    async def refresh_user(self, user, handler):
+        return {
+            "name": user.name,
+            "roles": self.refresh_roles,
+        }
+
+
+@pytest.mark.parametrize(
+    "authenticated_roles, refresh_roles",
+    [
+        ([{"name": "testrole-1", "users": "testuser-1"}], None),
+        ([{"name": "testrole-2", "users": "testuser-1"}], None),
+        ([{"name": "testrole-3", "users": "testuser-1"}], None),
+        ([{"name": "testrole-4", "users": "testuser-1"}], None),
+        ([{"name": "testrole-5", "users": "testuser-1"}], None),
+    ],
+)
+async def test_auth_managed_roles(app, user, role, authenticated_roles, refresh_roles):
+
+    authenticator = MockRolesAuthenticator(
+        parent=app,
+        authenticated_roles=authenticated_roles,
+    )
+    hub = MockHub(db_url='sqlite:///jupyterhub.sqlite')
+    user.roles.append(role)
+    app.db.commit()
+    before_roles = role.name
+
+    if authenticated_roles is None:
+        expected_authenticated_roles = before_roles
+    else:
+        expected_authenticated_roles = authenticated_roles
+
+    if authenticated_roles is None:
+        expected_authenticated_groups = before_roles
+    else:
+        expected_authenticated_groups = authenticated_roles
+    if refresh_roles is None:
+        expected_refresh_groups = expected_authenticated_groups
+    else:
+        expected_refresh_groups = refresh_roles
+
+    # Check if user gets auth-managed roles
+    with mock.patch.dict(app.tornado_settings, {"authenticator": authenticator}):
+        assert not app.db.dirty
+        all_roles = app.db.query(orm.Role).all()
+        user_roles = sorted(g.name for g in user.roles)
+        expected_authenticated_roles_names = getRoleNames(expected_authenticated_roles)
+        for name in expected_authenticated_roles_names:
+            assert name in user_roles
+            role = orm.Role.find(app.db, name)
+            app.db.delete(role)
+
+    # Check if roles are deleted after restart
+    await hub.initialize()
+    all_roles = app.db.query(orm.Role).all()
+    all_roles_names = sorted(g.name for g in all_roles)
+    for name in expected_authenticated_roles_names:
+        assert name not in all_roles_names
```

Comment:
I think this is what _shouldn't_ happen. The roles should not change across restart. If the roles are deleted on restart, then behavior is inconsistent. Hub restarts without configuration changes should be minimally disruptive (ideally not at all).

Reply:
Do I understand the problem correctly: when hub restarts users are not required to login again, thus the role data obtained during login on the Hub side may be incomplete or missing altogether?

Would it be preferable to:
- keep track of all active sessions, and on Hub restart force refresh `auth` state for all users  who were active when Hub got shut down, or
- make use of the existing `Authenticator.refresh_pre_spawn` config option which appears to have similar effect (on the level of individual users) as it calls `refresh_auth` with `force=True` when enabled; I imagine that `Authenticator.refresh_pre_spawn` could get a new `auto` value (coded as enum or `None`) which would mean "True if `manage_roles` is enabled".